delira/training/predictor.py

```diff
@@ -0,0 +1,277 @@
+        key_mapping : dict
+            a dictionary containing the mapping from the ``data_dict`` to 
+            the actual model's inputs.
+            E.g. if a model accepts one input named 'x' and the data_dict 
+            contains one entry named 'data' this argument would have to 
+            be ``{'x': 'data'}``
+        convert_batch_to_npy_fn : type, optional
+            function converting a batch-tensor to numpy, default: identity 
+            function
+        prepare_batch_fn : type, optional
+            function converting a batch-tensor to the framework specific 
+            tensor-type and pushing it to correct device, default: identity 
+            function
+
+        """
+
+        self._setup(model, key_mapping, convert_batch_to_npy_fn, 
+                    prepare_batch_fn, **kwargs)
+
+        self._tqdm_desc = "Test"
+
+    def _setup(self, network, key_mapping, convert_batch_to_npy_fn, 
+               prepare_batch_fn, **kwargs):
+        """
+
+        Parameters
+        ----------
+        network : :class:`AbstractNetwork`
+            the network to predict from
+        key_mapping : dict
+            a dictionary containing the mapping from the ``data_dict`` to 
+            the actual model's inputs.
+            E.g. if a model accepts one input named 'x' and the data_dict 
+            contains one entry named 'data' this argument would have to 
+            be ``{'x': 'data'}``
+        convert_batch_to_npy_fn : type
+            a callable function to convert tensors to numpy
+        prepare_batch_fn : type
+            function converting a batch-tensor to the framework specific
+            tensor-type and pushing it to correct device, default: identity
+            function
+
+        """
+        self.module = network
+        self.key_mapping = key_mapping
+        self._convert_batch_to_npy_fn = convert_batch_to_npy_fn
+        self._prepare_batch = prepare_batch_fn
+
+    def __call__(self, data: dict):
+        """
+        Method to call the class.
+        Returns the predictions corresponding to the given data 
+        obtained by the model
+
+        Parameters
+        ----------
+        data : dict
+            batch dictionary
+
+        Returns
+        -------
+        dict
+            predicted data
+        """
+        return self.predict(data)
+
+    def predict(self, data: dict):
+
+        data = self._prepare_batch(data)
+
+        mapped_data = {
+            k: data[v] for k, v in self.key_mapping.items()}
+
+        pred = self.module(
+                **mapped_data
+            )
+
+        # converts positional arguments and keyword arguments,
+        # but returns only keyword arguments, since positional
+        # arguments are not given.
+        return self._convert_batch_to_npy_fn(
+            **pred
+        )[1]
+
+    def predict_data_mgr(self, datamgr, batchsize=None, metrics={},
+                         metric_keys=None, verbose=False):
```

Comment:
Should we add the ability to compute metrics over the whole data_mgr? Instead of returning a list of batch_level metrics in addiction to all predictions. I think we had decided not to for training data_managers...
Also, description of what is returned is missing

Reply:
We had decided to do this in general but to postpone this, until we have a way to do this lazy and non-lazy, because for large datasets (e.g. imagenet) we might also break the lazy behavior with the validation/testset.